### PR TITLE
AHOYAPPS-531: Dynamic sender settings

### PR DESCRIPTION
**Pull request by timrozum**
https://issues.corp.twilio.com/browse/AHOYAPPS-531

1. Dynamically update frame rate based on number of participants when video codec is VP8 simulcast. 
1. Was trying to be surgical because this area of code is kind of hard to change and we were not expecting to spend a lot of time on this.

### CODE REVIEW

File: VideoApp/VideoApp/Video/LocalMediaController.m [hunk 1]
```diff
@@ -119,13 +121,22 @@ - (void)destroyLocalVideoTrack {
...
 
 - (BOOL)shouldMirrorLocalVideoView {
     return self.camera.shouldMirrorLocalVideoView;
 }
 
+- (void)checkVideoSenderSettings:(BOOL)isMultiparty {
+    if (SwiftToObjc.isVideoCodecVP8Simulcast && self.localParticipant != nil && isMultiparty != self.useMultipartyMedia) {
+        self.useMultipartyMedia = isMultiparty;
+        TVIEncodingParameters *encodingParameters = [[TVIEncodingParameters alloc] initWithAudioBitrate:0 videoBitrate:1800];
+        [self.localParticipant setEncodingParameters:encodingParameters];
```
Comment (timrozum) [inline VideoApp/VideoApp/Video/LocalMediaController.m:135]: This came over from MultiPartyExample but I'm thinking it isn't needed. Chris suggested only changing the frame rate dynamically. How is setting encoding parameters on local participant different from setting encoding parameters on connect options?

File: VideoApp/VideoApp/Video/VideoAppCameraSource.swift [hunk 2]
```diff
@@ -85,41 +95,20 @@ import TwilioVideo
...
-        let preferredFormat = selectVideoFormatBySize(captureDevice: captureDevice, targetSize: targetSize)
-        preferredFormat.frameRate = min(preferredFormat.frameRate, frameRate)
+        let preferredFormat = selectVideoFormat(captureDevice: captureDevice, isMultiparty: false)
         
         let cropDimensions: CMVideoDimensions
         
         if preferredFormat.dimensions.width > preferredFormat.dimensions.height {
             cropDimensions = CMVideoDimensions(
-                width: Int32(CGFloat(preferredFormat.dimensions.height) * cropRatio),
+                width: Int32(CGFloat(preferredFormat.dimensions.height) * videoCodec.cropRatio),
```
Comment (timrozum) [inline VideoApp/VideoApp/Video/VideoAppCameraSource.swift:105]: How come we only do this crop stuff here and not in other places like when the camera is switched from front to back?

File: VideoApp/VideoApp/Video/LocalMediaController.m [hunk 3]
```diff
@@ -119,13 +121,22 @@ - (void)destroyLocalVideoTrack {
...
 }
 
 - (BOOL)shouldMirrorLocalVideoView {
     return self.camera.shouldMirrorLocalVideoView;
 }
 
+- (void)checkVideoSenderSettings:(BOOL)isMultiparty {
+    if (SwiftToObjc.isVideoCodecVP8Simulcast && self.localParticipant != nil && isMultiparty != self.useMultipartyMedia) {
+        self.useMultipartyMedia = isMultiparty;
+        TVIEncodingParameters *encodingParameters = [[TVIEncodingParameters alloc] initWithAudioBitrate:0 videoBitrate:1800];
```
Comment (paynerc) [inline VideoApp/VideoApp/Video/LocalMediaController.m:134]: The is only changing the video bitrate to `kSimulcastVideoBitrate`, which is `1800`.

The code in the Multi-Party example app is doing a few other comparisons:

```
var bitrate = isMultiparty ? CaptureDeviceUtils.kMultipartyVideoBitrate : CaptureDeviceUtils.kOneToOneVideoBitrate
            bitrate = MultiPartyViewController.isSimulcast ? CaptureDeviceUtils.kSimulcastVideoBitrate : bitrate
            localParticipant.setEncodingParameters(EncodingParameters(audioBitrate: MultiPartyViewController.audioBitrate,
                                                                      videoBitrate: bitrate))
```

Where:
```
    static let kOneToOneFrameRate = UInt(24)
    static let kOneToOneVideoBitrate = UInt(1120)
    static let kMultipartyFrameRate = UInt(15)
    static let kMultipartyVideoBitrate = UInt(600)
    static let kSimulcastVideoBitrate = UInt(1800)
    static let kOneToOneVideoDimensions = CMVideoDimensions(width: 640, height: 480)
    static let kMultipartyVideoDimensions = CMVideoDimensions(width: 480, height: 360)
    // Produce 3 spatial layers ~ {1024x768, 512x384, 256x192}
    static let kSimulcastVideoDimensions = CMVideoDimensions(width: 1024, height: 768)
    static let kSimulcastVideoFrameRate = UInt(20)
```

So this only is working in the simulcast case. In the non-simulcast cases, `VP8` or `H.264` it should be able to go between `kOneToOneVideoBitrate` and `kMultipartyVideoBitrate`. In the cases of simulcast its always `kSimulcastVideoBitrate`.

File: VideoApp/VideoApp/Video/VideoAppCameraSource.swift [hunk 4]
```diff
@@ -172,3 +167,25 @@ extension VideoAppCameraSource: CameraSourceDelegate {
...
+            return CMVideoDimensions(width: 544, height: 480)
+        case .vp8Simulcast:
+            // 1024 x 768 squarish crop (1.25:1) on most iPhones. 1280 x 720 squarish crop (1.25:1) on the iPhone X
+            // and models that don't have 1024 x 768.
+            return CMVideoDimensions(width: 900, height: 720)
+        }
+    }
+    var cropRatio: CGFloat { CGFloat(targetSize.width) / CGFloat(targetSize.height) }
+    
+    func frameRate(isMultiparty: Bool) -> UInt {
```
Comment (paynerc) [inline VideoApp/VideoApp/Video/VideoAppCameraSource.swift:185]: The frame rates here don't seem to match with the constants declared in `MultiPartyViewController.swift`

From what I am seeing, If it's a 1:1 call and not simulcast it should be `24`. If its a multiparty call and not simulcast, it should be `15`. If its simulcast it should always be 20.

Comment (timrozum) [reply VideoApp/VideoApp/Video/VideoAppCameraSource.swift:185]: Yeah since I didn't really have comprehensive requirements I went with sort of a mashup in an effort to be surgical. So I kept the 20 and 24 values from video app and used the 15 value from MultiPartyExample since I thought multi party was the main thing we were trying to bring over. Not really sure what is desired but it's easy to change. What do you think?

Comment (timrozum) [reply VideoApp/VideoApp/Video/VideoAppCameraSource.swift:185]: I'll take another look at MultiPartyViewController.

Comment (timrozum) [reply VideoApp/VideoApp/Video/VideoAppCameraSource.swift:185]: I think I get it now. All of this configuration inside if statements and switches is so hard to follow especially after adding this PR. I want a config struct so bad.

Comment (timrozum) [reply VideoApp/VideoApp/Video/VideoAppCameraSource.swift:185]: I think this is fixed now. Good catch.

File: VideoApp/VideoApp/Video/LocalMediaController.m (hunk 3 above)
Comment (timrozum) [reply VideoApp/VideoApp/Video/LocalMediaController.m]: Yeah that for sure wasn't right. I think it's fixed now. It's still a little simpler than MultiPartyExample because it only gets inside of this `if` in the case of simulcast.

**Comment by ceaglest** [comment]
> Dynamically update frame rate based on number of participants when video codec is VP8 simulcast.

I will leave a more detailed review tomorrow, but I wanted to comment on this point. Simulcast is meant for multi-party Group Rooms, but dynamic sender settings should be used to optimize unicast for 1:1 vs multi-party scenarios in either Peer-to-Peer or Group Rooms.

**Comment by timrozum** [comment]
I think I may have read some instructions incorrectly from Chris earlier and made mistake to apply dynamic sender settings to simulcast instead of non-simulcast. It may be as easy to fix as switching a boolean. But I'm not feeling great about this PR. 

- I had assumed adjusting capture configuration would be pretty easy but it was not and so this was rushed because it was not team's main priority.
- I think I should have written down comprehensive capture requirements somewhere, maybe a spreadsheet, so that we could be sure everyone was on same page.
- There's too many if statements to follow. In the future I think we either need:
    - Completely separate capture logic for each capture method even if it means some duplicate code so that it is easy for us to understand and easy for customers to simply pick what implementation they want.
    - Or at least abstract things like frame rate, bitrate, and resolution to a config struct or something.
    - Or maybe we should keep MultiPartyExample.

If this PR is pretty close maybe we just continue forward but certainly open to other options.

**Comment by timrozum** [comment]
Talked offline and decided not to add dynamic sender settings to video app.
